task/control-plane-agent/src/mgs_sidecar.rs

```diff
@@ -166,19 +237,204 @@ impl MgsHandler {
 
         Ok(state)
     }
+
+    /// Unlocks the tech port if the challenge and response are compatible
+    fn unlock(
+        &mut self,
+        vid: u16,
+        challenge: UnlockChallenge,
+        response: UnlockResponse,
+        time_sec: u32,
+    ) -> Result<(), GwMonorailError> {
+        ringbuf_entry!(Trace::UnlockRequested(vid));
+
+        if time_sec > MAX_UNLOCK_TIME_SECS {
+            ringbuf_entry!(Trace::UnlockTimeTooLong(time_sec));
+            return Err(GwMonorailError::TimeIsTooLong);
+        }
+
+        let Some(vlan_index) = VLANS.iter().position(|v| v.vid == vid) else {
+            return Err(GwMonorailError::InvalidVLAN);
+        };
+
+        if VLANS[vlan_index].trusted {
+            return Err(GwMonorailError::AlreadyTrusted);
+        }
+
+        let Some((last_challenge, challenge_time)) =
+            core::mem::take(&mut self.last_challenge)
+        else {
+            ringbuf_entry!(Trace::NoChallenge);
+            return Err(GwMonorailError::UnlockAuthFailed);
+        };
+
+        if challenge != last_challenge {
+            ringbuf_entry!(Trace::WrongChallenge);
+            return Err(GwMonorailError::UnlockAuthFailed);
+        }
+
+        let now = sys_get_timer().now;
+        if now >= challenge_time + CHALLENGE_EXPIRATION_TIME_SECS * 1000 {
+            ringbuf_entry!(Trace::ChallengeExpired);
+            return Err(GwMonorailError::UnlockAuthFailed);
+        }
+
+        // Check that the response is valid for our challenge.  Right now,
+        // there's only one option, but we'll use a match to check for the
+        // possibility of new types being added.
+        match (challenge, response) {
+            (
+                UnlockChallenge::Trivial { timestamp: ts1 },
+                UnlockResponse::Trivial { timestamp: ts2 },
+            ) if ts1 == ts2 => Ok(()),
+            _ => Err(GwMonorailError::UnlockAuthFailed),
+        }?;
+        ringbuf_entry!(Trace::UnlockAuthSucceeded);
+
+        // Pick how long we'll trust things for
+        let now = sys_get_timer().now;
+        let trust_until = now + u64::from(time_sec) * 1000;
+
+        // Reconfigure the management network for arbitrary SP access
+        if let Err(e) = self.monorail.unlock_vlans(trust_until) {
+            ringbuf_entry!(Trace::MonorailUnlockFailed(e));
+            return Err(GwMonorailError::UnlockFailed);
+        }
+
+        // Reconfigure the net task to temporarily trust this port's VLAN
+        let net = task_net_api::Net::from(crate::NET.get_task_id());
+        if let Err(e) = net.trust_vlan(vid, trust_until) {
+            ringbuf_entry!(Trace::TrustVLanFailed(e));
+            return Err(GwMonorailError::UnlockFailed);
```

Comment:
(same question, right now these two failures are indistinguishable to the caller)

Reply:
Same thing here, we don't want to / can't easily expose internal `MonorailError` types to MGS.